engine/kv_engine.cpp

```diff
@@ -740,130 +710,303 @@ Status KVEngine::CheckConfigs(const Configs& configs) {
   if (s != Status::Ok) {
     return s;
   }
 
-  s = MaybeInitPendingBatchFile();
+  s = maybeInitBatchLogFile();
   if (s != Status::Ok) {
     return s;
   }
 
-  engine_thread_cache_[access_thread.id].batch_writing = true;
-  defer(engine_thread_cache_[access_thread.id].batch_writing = false;);
+  // Prevent generating snapshot newer than this WriteBatch
+  auto bw_token = version_controller_.GetBatchWriteToken();
+  // Prevent collection and nodes in double linked lists from being deleted
+  auto access_token = version_controller_.GetLocalSnapshotHolder();
+
+  // Lock all keys
+  std::vector<StringView> keys;
+  for (auto const& string_op : batch.StringOps()) {
+    keys.push_back(string_op.key);
+  }
+  for (auto const& sorted_op : batch.SortedOps()) {
+    keys.push_back(sorted_op.field);
+  }
+  for (auto const& hash_op : batch.HashOps()) {
+    keys.push_back(hash_op.field);
+  }
+  auto guard = hash_table_->RangeLock(keys);
 
-  std::set<SpinMutex*> spins_to_lock;
-  std::vector<BatchWriteHint> batch_hints(write_batch.Size());
-  std::vector<uint64_t> space_entry_offsets;
+  // Lookup keys, allocate space according to result.
+  std::vector<StringWriteArgs> string_args;
+  std::vector<SortedWriteArgs> sorted_args;
+  std::vector<HashWriteArgs> hash_args;
 
-  for (size_t i = 0; i < write_batch.Size(); i++) {
-    auto& kv = write_batch.kvs[i];
-    uint32_t requested_size;
-    if (kv.type == StringDataRecord) {
-      requested_size = kv.key.size() + kv.value.size() + sizeof(StringRecord);
-    } else if (kv.type == StringDeleteRecord) {
-      requested_size = kv.key.size() + sizeof(StringRecord);
-    } else {
-      GlobalLogger.Error("only support string type batch write\n");
-      return Status::NotSupported;
+  auto ReleaseResources = [&]() {
+    for (auto iter = hash_args.rbegin(); iter != hash_args.rend(); ++iter) {
+      pmem_allocator_->Free(iter->space);
     }
-    batch_hints[i].allocated_space = pmem_allocator_->Allocate(requested_size);
-    // No enough space for batch write
-    if (batch_hints[i].allocated_space.size == 0) {
+    for (auto iter = sorted_args.rbegin(); iter != sorted_args.rend(); ++iter) {
+      pmem_allocator_->Free(iter->space);
+    }
+    for (auto iter = string_args.rbegin(); iter != string_args.rend(); ++iter) {
+      pmem_allocator_->Free(iter->space);
+    }
+  };
+  defer(ReleaseResources());
+
+  // Prepare for Strings
+  for (auto const& string_op : batch.StringOps()) {
+    StringWriteArgs args;
+    args.Assign(string_op);
+    args.ts = bw_token.Timestamp();
+    args.res = lookupKey<true>(args.key, StringRecordType);
+    if (args.res.s != Status::Ok && args.res.s != Status::NotFound &&
+        args.res.s != Status::Outdated) {
+      return args.res.s;
+    }
+    if (args.op == WriteBatchImpl::Op::Delete && args.res.s != Status::Ok) {
+      // No need to do anything for delete a non-existing String
+      continue;
+    }
+    args.space = pmem_allocator_->Allocate(
```

Comment:
for delete, make sure string_op.value is empty? or if(delete) { RecordSize(key, ""); }